go/test/endtoend/shardedrecovery/sharded_recovery_test.go

```diff
@@ -0,0 +1,630 @@
+	assert.Nil(t, err)
+
+	err = localCluster.VtctlclientProcess.ExecuteCommand("MigrateServedTypes", "test_keyspace/0", "replica")
+	assert.Nil(t, err)
+
+	// then serve master from the split shards
+	err = localCluster.VtctlclientProcess.ExecuteCommand("MigrateServedTypes", "test_keyspace/0", "master")
+	assert.Nil(t, err)
+
+	// remove the original tablets in the original shard
+	removeTablets(t, []*cluster.Vttablet{master, replica1, rdOnly})
+
+	for _, tablet := range []*cluster.Vttablet{replica1, rdOnly} {
+		err = localCluster.VtctlclientProcess.ExecuteCommand("DeleteTablet", tablet.Alias)
+		assert.Nil(t, err)
+	}
+	err = localCluster.VtctlclientProcess.ExecuteCommand("DeleteTablet", "-allow_master", master.Alias)
+	assert.Nil(t, err)
+
+	// rebuild the serving graph, all mentions of the old shards should be gone
+	err = localCluster.VtctlclientProcess.ExecuteCommand("RebuildKeyspaceGraph", "test_keyspace")
+	assert.Nil(t, err)
+
+	// delete the original shard
+	err = localCluster.VtctlclientProcess.ExecuteCommand("DeleteShard", "test_keyspace/0")
+	assert.Nil(t, err)
+
+	// now bring up the recovery keyspace and a tablet, letting it restore from backup.
+	restoreTablet(t, replica2, recoveryKS, "0")
+
+	// check the new replica does not have the data
+	checkData(t, replica2, 2)
+
+	vtgateInstance := localCluster.GetVtgateInstance()
+	vtgateInstance.TabletTypesToWait = "REPLICA"
+	err = vtgateInstance.Setup()
+	localCluster.VtgateGrpcPort = vtgateInstance.GrpcPort
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", keyspaceName, shard1Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", keyspaceName, shard1Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", recoveryKS, shardName), 1)
+	require.Nil(t, err)
+
+	// Build vtgate grpc connection
+	// check that vtgate doesn't route queries to new tablet
+	grpcAddress := fmt.Sprintf("%s:%d", localCluster.Hostname, localCluster.VtgateGrpcPort)
+	vtgateConn, err := vtgateconn.Dial(context.Background(), grpcAddress)
+	assert.Nil(t, err)
+	session := vtgateConn.Session("@replica", nil)
+	verifyQueriesUsingVtgate(t, session, "select count(*) from vt_insert_test", "INT64(3)")
+
+	// check that new tablet is accessible by using ks.table
+	verifyQueriesUsingVtgate(t, session, "select count(*) from recovery_keyspace.vt_insert_test", "INT64(2)")
+
+	// check that new tablet is accessible with 'use ks'
+	executeQueriesUsingVtgate(t, session, "use recovery_keyspace@replica")
+	verifyQueriesUsingVtgate(t, session, "select count(*) from vt_insert_test", "INT64(2)")
+
```

Comment:
We should also test the case that we weren't able to test in python:
https://github.com/vitessio/vitess/blob/master/test/sharded_recovery.py#L382

Reply:
done